Rich command: Context + Restore

cinqmilleans opened the pull request:
**Resolved / Related Issues**
This pr adds the first and main context used by actions. A context is an observable service that easily provides actions with properties that are difficult to know. To observe properties of the current page, you must notify the properties when it changes or when a parent or ancestor object changes.

* Added ContentPageContext. The data is always from the current page. Selected items are easily notified (whether it's a selection or a page change).
* Added RestoreRecycleBin and RestoreAllRecycleBin commands.
* Improved EmptyRecycleBin with context.

**Validation**
How did you test these changes?
- [x] Built and ran the app
- [ ] Tested the changes for accessibility

d2dyno1 commented on src/Files.App/Contexts/ContentPage/IContentPageContext.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace Files.App.Contexts
+{
+	public interface IContentPageContext : INotifyPropertyChanged
```
What is this interface for?

cinqmilleans replied:
This interface is a context service which gives the active content of the page (The type of page, the selected elements, ...).
The properties of this context are useful to most commands and must be observable for the command to notify its state changes.

If I want to know the selected items, currently I have to observe ToolbarViewModel, but this instance itself can change
without notification if I switch tabs. It is therefore necessary to observer several "floors" of an object and update each time the floor changes. The context makes its information easy to obtain and notifies of changes.

d2dyno1 replied:
I guess it's okay, but on the other hand it bothers me because we're just further spreading the use of IShellPage...

cinqmilleans replied:
What do you mean ?

yaira2 replied:
@d2dyno1 is this good to go?

QuaintMako commented on src/Files.App/Actions/FileSystem/EmptyRecycleBinAction.cs:
```diff
@@ -1,19 +1,52 @@
+		{
+			get
+			{
+				if (context.PageType is ContentPageTypes.RecycleBin)
+					return context.HasItem;
+				return RecycleBinHelpers.RecycleBinHasItems();
```
Can we just use a simple `RecycleBinHelpers.RecycleBinHasItems()` here? If you did it, I suppose it's because the `context.hasItem` is lighter to execute?

cinqmilleans replied:
Yes, context.HasItem is 5-10 times faster. But this is a trick that can only be used if the RecycleBin is displayed in the current page.

QuaintMako commented on src/Files.App/Actions/FileSystem/EmptyRecycleBinAction.cs:
```diff
@@ -1,19 +1,52 @@
+
+		public EmptyRecycleBinAction()
+		{
+			context.PropertyChanged += Context_PropertyChanged;
```
No need to unsubscribe somewhere?

QuaintMako commented on src/Files.App/Actions/FileSystem/RestoreAllRecycleBinAction.cs:
```diff
@@ -0,0 +1,53 @@
+		{
+			get
+			{
+				if (context.PageType is ContentPageTypes.RecycleBin)
+					return context.HasItem;
+				return RecycleBinHelpers.RecycleBinHasItems();
```
Same as for EmptyRecycleBin

QuaintMako commented on src/Files.App/Actions/FileSystem/RestoreAllRecycleBinAction.cs:
```diff
@@ -0,0 +1,53 @@
+
+		public RestoreAllRecycleBinAction()
+		{
+			context.PropertyChanged += Context_PropertyChanged;
```
Same as for EmptyRecycleBin

cinqmilleans replied:
Same as for EmptyRecycleBin

QuaintMako commented on src/Files.App/Actions/FileSystem/RestoreRecycleBinAction.cs:
```diff
@@ -0,0 +1,46 @@
+
+		public RestoreRecycleBinAction()
+		{
+			context.PropertyChanged += Context_PropertyChanged;
```
Same has EmptyRecycleBin

cinqmilleans replied:
Same as for EmptyRecycleBin

cinqmilleans commented on src/Files.App/Commands/Manager/CommandManager.cs:
```diff
@@ -121,6 +122,9 @@ private class ActionCommand : ObservableObject, IRichCommand
 			private readonly Lazy<ColoredIcon?> coloredIcon;
 			public ColoredIcon? ColoredIcon => coloredIcon.Value;
```
It's done.

cinqmilleans commented:
I updated to resolve an incompatibility with the latest commits. This forced me to integrate the management of OpacityIcon

* RichGlyph can be of type OpacityIcon
* ColoredIcon is removed.
* Restore(All) uses OpacityIcon.
* Added ContextMenuFlyoutItemViewModelBuilder. This new class avoids having to read the properties of the command when it is not displayed in the menu, to improve performance.

When all the commands have been converted, it will be possible to replace ContextMenuFlyoutItemViewModel with an interface and 3 implementations to improve code and performance.

Known bug: In the context menu, the Restore entry of a RecycleBin item has no icon. However, the element does have an OpacityIcon content. I don't think the bug is related to this code.

yaira2 commented:
> Known bug: In the context menu, the Restore entry of a RecycleBin item has no icon. However, the element does have an OpacityIcon content. I don't think the bug is related to this code.

I noticed this as well, I think we can leave as is for now (same issue in `main`)

yaira2 commented on src/Files.App/App.xaml:
```diff
@@ -6,8 +6,6 @@
 		<ResourceDictionary>
 			<FontFamily x:Key="CustomGlyph">/Assets/Custom Glyphs/Custom-Glyphs.ttf#Custom-Glyphs</FontFamily>
-			<FontFamily x:Key="RecycleBinIcons">/Assets/Custom Glyphs/recyclebin-fabric-glyph.ttf#Fabric MDL2 Assets</FontFamily>
-			<FontFamily x:Key="ColoredIconFontFamily">/Assets/Custom Glyphs/Colored-Icons.ttf#Untitled1</FontFamily>
```
@winston-de 😂

yaira2 reviewed (approved):
LGTM
